nupic/research/frameworks/dendrites/modules/boosted_dendrites.py

```diff
@@ -0,0 +1,195 @@
+    apply_boosted = BoostedDendrites(num_units, num_segments, ...)
+
+    # Run forward pass given input x and context.
+    y = net(x)
+    dendrite_activations = dendrite_segments(context)
+
+    # Apply boosted activations; duty cycles are automatically updated in training mode.
+    y_new, winning_indices = apply_boosted(y, dendrite_activations)
+
+    # Update boost_strength after each epoch.
+    segments_booster.update_boost_strength()
+    ```
+
+    :param num_units: number of units i.e. neurons;
+                      each unit has it's own set of dendrite segments
+    :param num_segments: number of dendrite segments per unit
+    :param boost_strength: initial boost strength; the higher the value, the more
+                           scarcely winning segments will be promoted
+    :param boost_strength_factor: the boost_strength decay factor; the user calls
+                                  `update_boost_strength` to multiply this factor by the
+                                  current boost strength.
+    :param duty_cycle_period: the size of the rolling window to calculate the frequency
+                              of times each segments wins for a given unit.
+
+    .. _KWinners:
+        https://github.com/numenta/nupic.torch/blob/master/nupic/torch/modules/k_winners.py
+    """
+
+    def __init__(
+        self,
+        num_units,
+        num_segments,
+        boost_strength=1.0,
+        boost_strength_factor=1.0,
+        duty_cycle_period=1000,
+    ):
+        super().__init__()
+        assert boost_strength >= 0.0
+        assert 0.0 <= boost_strength_factor <= 1.0
+
+        self.num_units = num_units
+        self.num_segments = num_segments
+        self.boost_strength_factor = boost_strength_factor
+        self.duty_cycle_period = duty_cycle_period
+
+        boost_strength = torch.tensor(boost_strength, dtype=torch.float)
+        duty_cycles = torch.zeros((num_units, num_segments))
+
+        self.register_buffer("duty_cycles", duty_cycles)
+        self.register_buffer("boost_strength", boost_strength)
+        self.register_buffer("learning_iterations", torch.tensor(0))
+
+    def extra_repr(self):
+        return (
+            f"num_units={self.num_units}, "
+            f"num_segments={self.num_segments}, "
+            f"boost_strength={self.boost_strength}, "
+            f"boost_strength_factor={self.boost_strength_factor}, "
+            f"duty_cycle_period={self.duty_cycle_period}"
+        )
+
+    def boost_activations(self, dendrite_activations):
+        """
+        Boost dendrite activations according to the duty cycles of the winning segments.
+
+        :param dendrite_activations: output from a DendriteSegments module;
+                                     shape batch_size x num_units x num_segments
+        """
+        boosted = dendrite_activations.detach()
+        boosted *= torch.exp(-self.boost_strength * self.duty_cycles)
```

Comment:
(This relates to the comment above) `boosted` will approach 0 as `boost_strength` approaches infinity. I find that to be unintuitive.

Reply:
When `boost_strength` is 0, then `boosted` will be 1 (exp(0) = 1). Otherwise, boosted will be somewhere between 0 and 1. A higher duty cycle will make boosted closer to 0 than a lower duty cycle. For instance,
```
boost_strength = 0.3
np.exp(-boost_strength * 4) # lower duty cycle
>>> 0.30
np.exp(-boost_strength * 5) # lower duty cycle
>>> 0.22
```

Thus the units with the lower duty cycle have greater multipliers applied to their activations. This effect is amplified for higher boost strengths earlier on in training. Say `boost_strength = 1`, then in the example from before
```
boost_strength = 1
np.exp(-boost_strength * 4)
>>> 0.018
np.exp(-boost_strength * 5)
>>> 0.006
```
Now, the high duty cycle unit must have an activation three times greater than the lower duty cycle to win. Whereas before (`boost_strength = 0.3`) it only needed to be about 50% higher to win.